File: src/main/dart/g0001_0100/s0031_next_permutation/Solution.dart
// #Medium #Top_100_Liked_Questions #Array #Two_Pointers #Big_O_Time_O(n)_Space_O(1)
// #2024_10_03_Time_315_ms_(94.12%)_Space_148.7_MB_(47.06%)

class Solution {
  void nextPermutation(List<int> nums) {
    if (nums.length <= 1) {
      return;
    }
    int i = nums.length - 2;

    // Find the first index `i` where nums[i] < nums[i + 1]
    while (i >= 0 && nums[i] >= nums[i + 1]) {
      i--;
    }

    // If such an index exists, find the next larger element to swap with
    if (i >= 0) {
      int j = nums.length - 1;
      while (nums[j] <= nums[i]) {
        j--;
      }
      _swap(nums, i, j);
    }

    // Reverse the part of the array after index `i`
    _reverse(nums, i + 1, nums.length - 1);
  }

  // Helper function to swap two elements in the array
  void _swap(List<int> nums, int i, int j) {
    int temp = nums[i];
    nums[i] = nums[j];
    nums[j] = temp;
  }

  // Helper function to reverse the elements of the array between indices i and j
  void _reverse(List<int> nums, int i, int j) {
    while (i < j) {
      _swap(nums, i++, j--);
    }
  }
}

File: src/main/c/g0001_0100/s0031_next_permutation/Solution.c
// #Medium #Top_100_Liked_Questions #Array #Two_Pointers #Big_O_Time_O(n)_Space_O(1)
// #2024_10_29_Time_0_ms_(100.00%)_Space_10.6_MB_(91.02%)

#include <stdio.h>

// Function to swap two elements in the array
void swap(int* nums, int i, int j) {
    int temp = nums[i];
    nums[i] = nums[j];
    nums[j] = temp;
}

// Function to reverse elements in the array from index i to j
void reverse(int* nums, int i, int j) {
    while (i < j) {
        swap(nums, i++, j--);
    }
}

// Function to find the next permutation
void nextPermutation(int* nums, int numsSize) {
    if (nums == NULL || numsSize <= 1) {
        return;
    }

    int i = numsSize - 2;
    // Find the first decreasing element from the end
    while (i >= 0 && nums[i] >= nums[i + 1]) {
        i--;
    }

    if (i >= 0) {
        int j = numsSize - 1;
        // Find the element just larger than nums[i]
        while (nums[j] <= nums[i]) {
            j--;
        }
        swap(nums, i, j);
    }

    // Reverse the elements after the found index to get the next smallest lexicographic order
    reverse(nums, i + 1, numsSize - 1);
}
